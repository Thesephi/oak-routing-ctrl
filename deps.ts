export { join } from "jsr:@std/path@^0.225.2";

export { Router } from "jsr:@oak/oak@^16.1.0";

export type {
  Application,
  Context,
  Next,
  RouteContext,
} from "jsr:@oak/oak@^16.1.0";

import {
  type ResponseConfig,
  type RouteConfig,
} from "npm:@asteasolutions/zod-to-openapi@^7.1.1";

export type OakOpenApiSpec =
  & Omit<RouteConfig, "method" | "path" | "responses">
  & {
    responses?: {
      [statusCode: string]: ResponseConfig;
    };
  };

export { type ResponseConfig, type RouteConfig };

export {
  extendZodWithOpenApi,
  OpenApiGeneratorV3,
  OpenAPIRegistry,
  type ZodRequestBody,
} from "npm:@asteasolutions/zod-to-openapi@^7.1.1";

export { type OpenAPIObjectConfig } from "npm:@asteasolutions/zod-to-openapi@^7.1.1/dist/v3.0/openapi-generator";

export { z } from "npm:zod@^3.23.8"; // must export from npm: instead from deno.land to be compatible with zod-to-openapi
