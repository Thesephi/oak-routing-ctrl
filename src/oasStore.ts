import { type OakOpenApiSpec, type RouteConfig } from "../deps.ts";
import { SupportedVerb } from "./Store.ts";
import { debug } from "./utils/logger.ts";

// export type OasRouteConfig = {
//   method: SupportedVerb;
//   path: string;
//   request: {
//     params?: object; // RouteParameter from Zod
//     body?: ZodRequestBody;
//     query?: object; // RouteParameter from Zod
//   };
//   responses?: {
//     [statusCode: string]: ResponseConfig;
//   };
//   // @TODO cookies & headers
// };

// fnName|method|path => OasRouteConfig
export const oasStore: Map<string, RouteConfig> = new Map();

const getRouteId = (
  fnName: string,
  method: SupportedVerb,
  path: string,
) => `${fnName}|${method}|${path}`;

/**
 * input: `/some/:foo/and/:bar`
 * output: `/some/{foo}/and/{bar}`
 */
const getOasCompatPath = (path: string) =>
  path.replace(
    /\/:([a-zA-Z0-9]*)/g,
    "/{$1}",
  );

/**
 * updates OpenApiSpec for a given API resource
 * identified (keyed) by the tuple (fnName + method + path);
 * @NOTE this function can be called multiple times
 * to accumulatively update the same API resources,
 * e.g. replace a "short path" with a "full path"
 */
export const updateOas = (
  fnName: string,
  method: SupportedVerb,
  path: string,
  specs?: Partial<OakOpenApiSpec>,
): void => {
  // if developer doesn't supply `specs` then bail early
  // because we don't want "documentation without consent"
  if (!specs) return;

  const oasRouteIdentifier = getRouteId(fnName, method, path);

  const oasPath = getOasCompatPath(path);

  const existing = oasStore.get(oasRouteIdentifier) || {
    method,
    path: oasPath,
    request: {},
    responses: {},
  };

  const updated = {
    ...existing,
    method,
    path: oasPath,
    request: {
      ...existing.request,
      ...specs?.request,
    },
    responses: {
      ...existing.responses,
      ...specs?.responses,
    },
  };

  debug(`OpenApiSpec: recording for [${method}] ${path}`);

  oasStore.set(oasRouteIdentifier, updated);
};

export const patchOasPath = (
  fnName: string,
  method: SupportedVerb,
  path: string,
) => {
  oasStore.forEach((storedSpecs, routeId) => {
    const [storedFnName, storedMethod, storedPath] = routeId.split("|");
    if (
      fnName === storedFnName &&
      method === storedMethod &&
      path.length > storedPath.length &&
      path.endsWith(storedPath)
    ) {
      debug(`OpenApiSpec: patching ${storedSpecs.path} to ${path}`);
      storedSpecs.path = getOasCompatPath(path);
    }
  });
};
