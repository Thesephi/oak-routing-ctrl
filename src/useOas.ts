import { Application } from "@oak/oak";
import { oasStore } from "./oasStore.ts";
import {
  OpenApiGeneratorV3,
  OpenAPIRegistry,
  type RouteConfig,
} from "@asteasolutions/zod-to-openapi";
import { type OpenAPIObjectConfig } from "@asteasolutions/zod-to-openapi/dist/v3.0/openapi-generator";
import { debug } from "./utils/logger.ts";
import { inspect } from "./utils/inspect.ts";

const defaultOasJsonServingPath = "/oas.json";
const defaultOasUiServingPath = "/swagger";
const defaultOasUiTemplate = `
<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <script type="module" src="https://unpkg.com/rapidoc/dist/rapidoc-min.js"></script>
  </head>
  <body>
    <rapi-doc spec-url="/oas.json" render-style="view" theme="dark"></rapi-doc>
  </body>
</html>
`;

const registry = new OpenAPIRegistry();

/**
 * interface for an object used to configure how the Open API Spec
 * is generated (e.g. `/oas.json`)
 */
export type UseOasConfig = Partial<OpenAPIObjectConfig> & {
  jsonPath?: string;
  uiPath?: string;
  uiTemplate?: string;
};

type UseOas = (
  app: Application,
  cfg?: UseOasConfig,
) => void;

const _useOas: UseOas = (
  app,
  cfg = {},
) => {
  const {
    jsonPath = defaultOasJsonServingPath,
    uiPath = defaultOasUiServingPath,
    uiTemplate = defaultOasUiTemplate,
    ...oasCfg
  } = cfg;

  oasStore.forEach((rc, fnName) => {
    if ("path" in rc && "method" in rc && "responses" in rc) {
      registry.registerPath(rc as RouteConfig);
    } else {
      debug(
        `WARNING: OAS RouteConfig for '${fnName}' lacks either 'path', 'method', or 'responses'`,
      );
    }
  });

  const generator = new OpenApiGeneratorV3(registry.definitions);
  const apiDoc = generator.generateDocument({
    openapi: "3.0.0",
    info: {
      version: "1.0.0",
      title: "My API",
      description: "This is the API",
    },
    servers: [{ url: "/" }],
    ...oasCfg,
  });

  debug("OpenApiSpec", inspect(apiDoc, { depth: 10 }));

  app.use(async (ctx, next) => {
    if (ctx.request.url.pathname === jsonPath) {
      ctx.response.body = apiDoc;
    } else if (ctx.request.url.pathname === uiPath) {
      ctx.response.body = uiTemplate;
    }
    await next();
  });
};

/**
 * helper method to enable Open API Spec for the routes
 * declared with oak-routing-ctrl decorators
 * @param {Application} app the {@linkcode Application} instance from `@oak/oak`
 * @param {UseOasConfig} [cfg] optional configuration object to
 * finetune the Open API spec documentation
 */
export const useOas = (
  app: Application,
  cfg: UseOasConfig = {},
): void => {
  try {
    _useOas(app, cfg);
  } catch (e) {
    debug(
      "unable to complete OpenApiSpec initialization:",
      (e as Error).message,
    );
  }
};

export const _internal = { registry };
