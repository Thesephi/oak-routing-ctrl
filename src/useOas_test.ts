import {
  assertEquals,
  assertSnapshot,
  createMockApp,
  stub,
} from "../dev_deps.ts";
import { OpenApiGeneratorV3, z } from "../deps.ts";
import { updateOas } from "./oasStore.ts";
import { _internal, useOas } from "./useOas.ts";

Deno.test("useOas when _useOas throws", () => {
  const mock = stub(_internal, "_useOas", () => {
    throw new Error("nope");
  });
  useOas(createMockApp());
  mock.restore();
});

Deno.test("useOas with empty definitions", () => {
  useOas(createMockApp());
  assertEquals(_internal.registry.definitions, []);
});

Deno.test("useOas standard behavior", async (t) => {
  const fnName = "doSomething";
  const method = "post";
  const path = "/hello/:name";

  updateOas(fnName, method, path, {
    request: {
      params: z.object({ name: z.string() }),
    },
  });

  const oasConfig = {
    uiPath: "/my/swagger",
    jsonPath: "/my/swagger/json",
    uiTemplate: "<html>mock</html>",
    openapi: "3.0.3",
    info: {
      version: "0.1.0",
      title: "mock API",
      description: "this is a mock API",
    },
    servers: [{ url: "/mock/" }],
  };

  useOas(createMockApp(), oasConfig);

  const generator = new OpenApiGeneratorV3(_internal.registry.definitions);
  const apiDoc = generator.generateDocument(oasConfig);

  await assertSnapshot(t, apiDoc);
});
