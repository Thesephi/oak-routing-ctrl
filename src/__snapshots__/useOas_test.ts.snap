export const snapshot = {};

snapshot[`useOas standard behavior 1`] = `
{
  components: {
    parameters: {},
    schemas: {},
  },
  info: {
    description: "this is a mock API",
    title: "mock API",
    version: "0.1.0",
  },
  jsonPath: "/my/swagger/json",
  openapi: "3.0.3",
  paths: {
    "/hello/{name}": {
      post: {
        parameters: [
          {
            in: "path",
            name: "name",
            required: true,
            schema: {
              type: "string",
            },
          },
        ],
        responses: {},
      },
    },
  },
  servers: [
    {
      url: "/mock/",
    },
  ],
  uiPath: "/my/swagger",
  uiTemplate: "<html>mock</html>",
}
`;
